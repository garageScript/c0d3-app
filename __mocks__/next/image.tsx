import React from 'react'
import type { ImageProps } from 'next/image'
import _ from 'lodash'

const Image = (props: ImageProps) => {
  /*not all Image properties are allowed in <img/> tag
  ObjectFit causes warnings, but otherwise lessons/error images won't keep their dimensions 
  see https://developer.mozilla.org/en-US/docs/Web/CSS/object-fit*/
  const copy = _.omit(props, 'objectFit')
  return <img {...copy} />
}

export default Image
