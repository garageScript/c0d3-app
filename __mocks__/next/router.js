const push = jest.fn()
const query = {}
export const useRouter = jest.fn().mockImplementation(() => {
  return {
    route: '/',
    pathname: '/',
    query,
    asPath: '/',
    push: push,
    replace: async () => true,
    reload: () => null,
    back: () => null
  }
})
