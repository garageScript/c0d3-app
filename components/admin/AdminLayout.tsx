import React from 'react'
import LoadingSpinner from '../../components/LoadingSpinner'
import Layout from '../../components/Layout'
import _ from 'lodash'
import { GetAppProps } from '../../graphql'
import Error, { StatusCode } from '../../components/Error'
import { useRouter } from 'next/router'
export const AdminLayout: React.FC<GetAppProps & { title?: string }> = ({
  data,
  children,
  title
}) => {
  const router = useRouter()
  const { loading, error, session } = data

  if (loading) {
    return <LoadingSpinner />
  }

  if (error) {
    return <Error code={StatusCode.INTERNAL_SERVER_ERROR} />
  }

  if (!session) router.push('/login')

  const isAdmin = _.get(session, 'user.isAdmin', false) as boolean

  if (!isAdmin) {
    return (
      <Layout title={title}>
        <h1>You must be admin to access this page</h1>
      </Layout>
    )
  }

  return <Layout>{children}</Layout>
}
