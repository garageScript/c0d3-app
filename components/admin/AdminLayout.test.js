import React from 'react'
import { render, waitFor } from '@testing-library/react'
import '@testing-library/jest-dom'
import { MockedProvider } from '@apollo/client/testing'
import { AdminLayout } from './AdminLayout'
import { useRouter } from 'next/router'
import expectLoading from '../../__tests__/utils/expectLoading'

describe('AdminLayout test', () => {
  const { push } = useRouter()
  test('Should return loading spinner when loading', async () => {
    expectLoading(<AdminLayout data={{ loading: true }} />)
  })
  test('Should return Error on error', async () => {
    const { findByText } = render(
      <MockedProvider addTypename={false}>
        <AdminLayout data={{ error: true }} />
      </MockedProvider>
    )
    const element = await findByText(/500/i)
    expect(element).toBeTruthy()
  })
  test('Should redirect to login page if no session', async () => {
    render(
      <MockedProvider addTypename={false}>
        <AdminLayout data={{ data: {} }} />
      </MockedProvider>
    )
    await waitFor(() => expect(push).toBeCalledWith('/login'))
  })
  test('Should render children', async () => {
    const { container } = render(
      <MockedProvider addTypename={false}>
        <AdminLayout
          data={{
            session: {
              user: {
                isAdmin: true
              }
            }
          }}
        >
          <h1>Test title</h1>
        </AdminLayout>
      </MockedProvider>
    )
    expect(container).toMatchSnapshot()
  })
})
