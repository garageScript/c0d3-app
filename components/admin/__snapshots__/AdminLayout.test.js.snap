// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminLayout test Should render children 1`] = `
<div>
  <nav
    class="navbar navbar-expand-lg navbar-light justify-content-between bg-white"
  >
    <div
      class="container"
    >
      <a
        class="navbar-brand text-primary font-weight-bold"
        href="/"
      >
        C0D3
      </a>
      <div
        id="navbarNav"
      >
        <div
          class="navbar-nav collapse navbar-collapse"
        >
          <div
            class="navbar-nav collapse navbar-collapse"
          >
            <a
              class="nav-item nav-link"
              href="/curriculum"
            >
              Curriculum
            </a>
            <a
              class="nav-item nav-link"
              href="https://github.com/garageScript/c0d3-app"
              rel="noopener noreferrer"
              target="_blank"
            >
              Repo
            </a>
            <a
              class="nav-item nav-link"
              href="https://www.notion.so/Table-of-Contents-a83980f81560429faca3821a9af8a5e2"
              rel="noopener noreferrer"
              target="_blank"
            >
              Journey
            </a>
            <a
              class="nav-item nav-link"
              href="https://chat.c0d3.com"
              rel="noopener noreferrer"
              target="_blank"
            >
              Help
            </a>
            <a
              class="nav-item nav-link"
              href="/contributors"
            >
              Contributors
            </a>
          </div>
        </div>
      </div>
      <div>
        <a
          class="btn btn-secondary border mr-3"
          href="/login"
        >
          Login
        </a>
        <a
          class="btn btn-secondary border mr-3"
          href="/signup"
        >
          Signup
        </a>
      </div>
    </div>
  </nav>
  <div
    class="container"
  >
    <h1>
      Test title
    </h1>
  </div>
  <footer
    class="mt-5 font-weight-light text-center pb-5 text-muted"
  >
    Â© Copyright 2020 GarageScript
    <br />
    All Rights Reserved
  </footer>
</div>
`;
